// @varlet/ui/es/date-picker/props.mjs
import { defineListenerProp } from "../utils/components.mjs";
const MONTH_LIST = ["01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "11", "12"];
const WEEK_HEADER = ["0", "1", "2", "3", "4", "5", "6"];
const props = {
  modelValue: [String, Array],
  type: {
    type: String,
    default: "date"
  },
  hint: String,
  allowedDates: Function,
  color: String,
  /** @deprecated Use titleColor to instead. */
  headerColor: String,
  titleColor: String,
  elevation: {
    type: [Boolean, Number, String],
    default: false
  },
  buttonElevation: {
    type: [Boolean, Number, String],
    default: true
  },
  firstDayOfWeek: {
    type: [String, Number],
    default: 0
  },
  min: String,
  max: String,
  showCurrent: {
    type: Boolean,
    default: true
  },
  readonly: Boolean,
  multiple: Boolean,
  range: Boolean,
  touchable: {
    type: Boolean,
    default: true
  },
  onPreview: defineListenerProp(),
  onChange: defineListenerProp(),
  "onUpdate:modelValue": defineListenerProp()
};
export {
  MONTH_LIST,
  WEEK_HEADER,
  props
};
